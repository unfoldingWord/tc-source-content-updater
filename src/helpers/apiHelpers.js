import {delay} from './utils';
import {SORT, STAGE, SUBJECT} from '../index';

const request = require('request');
const DOOR43_CATALOG = `Door43-Catalog`;

/**
 * does http request and returns the response data parsed from JSON
 * @param {string} url
 * @param {number} retries
 * @return {Promise<{Object}>}
 */
export async function makeJsonRequestDetailed(url, retries=5) {
  let result_;
  for (let i = 1; i <= retries; i++) {
    result_ = null;
    try {
      result_ = await new Promise((resolve, reject) => {
        request(url, function(error, response, body) {
          if (error)
            reject(error);
          else if (response.statusCode === 200) {
            let result = body;
            try {
              result = JSON.parse(body);
            } catch (e) {
              reject(e);
            }
            resolve({result, response, body});
          } else {
            reject(`fetch error ${response.statusCode}`);
          }
        });
      });
    } catch (e) {
      if (i >= retries) {
        console.warn('makeJsonRequestDetailed() - error getting manifest data', e);
        throw e;
      }
      result_ = null;
      await delay(500);
      console.log(`makeJsonRequestDetailed() - retry ${i+1} getting manifest data, last error`, e);
    }

    if (result_) {
      break;
    }
  }
  return result_;
}

/**
 * searching subjects
 * @param {array} subjects
 * @param {string} owner
 * @param {number} retries
 * @return {Promise<*>}
 */
async function searchSubjects(subjects, owner, retries=3) {
  const subjectParam = encodeURI(subjects.join(','));
  let fetchUrl = `https://git.door43.org/api/catalog/v3/search?subject=${subjectParam}`;
  if (owner) {
    fetchUrl += fetchUrl + `&owner=${owner}`;
  }
  const {result} = await makeJsonRequestDetailed(fetchUrl, retries);
  return result;
}

/**
 * get published catalog using catalog next v3
 * @return {Promise<*[]>}
 */
export async function getOldCatalogReleases() {
  const released = [];
  const owner = null; // get all owners
  const subjectList = ['Bible', 'Aligned Bible', 'Greek New Testament', 'Hebrew Old Testament', 'Translation Words', 'TSV Translation Notes', 'Translation Academy'];
  // const subjectList = ['Bible', 'Testament', 'Translation Words', 'TSV Translation Notes', 'Translation Academy'];

  try {
    // for (const subject of subjectList) {
    const result = await searchSubjects(subjectList, owner, 5);
    let repos = 0;
    const languages = result && result.languages || [];
    for (const language of languages) {
        const languageId = language.identifier;
        const resources = language.resources || [];
        for (const resource of resources) {
          resource.languageId = languageId;
          resource.resourceId = resource.identifier;
          resource.foundInCatalog = 'OLD';
          resource.full_name = resource.full_name || `${resource.owner}/${resource.repo}`;
          resource.checking_level = resource.checking && resource.checking.checking_level;
          released.push(resource);
          repos++;
        }
      }
      console.log(`has ${repos} items`);
    // }
    console.log(`released catalog has ${released.length} items`);
  } catch (e) {
    console.error('getCatalog() - error getting catalog', e);
    return [];
  }

  return released;
}

/**
 * get published catalog
 * @return {Promise<*[]>}
 */
export async function getCatalog() {
  const catalogReleases = await getOldCatalogReleases();
  const searchParams = {
    subject: SUBJECT.ALL_TC_RESOURCES,
    stage: STAGE.PROD,
  };
  console.log(`found ${catalogReleases.length} items in old catalog`);
  const newCatalogReleases = await searchCatalogNext(searchParams);
  console.log(`found ${newCatalogReleases.length} items in new catalog`);
  // merge catalogs together - catalog new takes precident
  for (const item of newCatalogReleases) {
    const index = catalogReleases.findIndex(oldItem => (item.full_name === oldItem.full_name));
    if (index >= 0) {
      catalogReleases[index] = item; // overwrite item in old catalog
      catalogReleases[index].foundInCatalog = 'NEW+OLD';
    } else {
      catalogReleases.push(item); // add unique item
    }
  }
  console.log(`now ${catalogReleases.length} items in merged catalog`);

  return catalogReleases;
}

/**
 * add parameter to url string
 * @param {*} value
 * @param {object} parameters
 * @param {string} tag
 * @return {*}
 */
function addUrlParameter(value, parameters, tag) {
  if (value) {
    if (parameters) { // see if we need separator
      parameters += '&';
    }
    parameters += `${tag}=${encodeURIComponent(value)}`;
  }
  return parameters;
}

/**
 * filter for supported repos
 * @param {array} resources
 * @return {*[]}
 */
function getCompatibleResourceList(resources) {
  const supported = [];
  for (const item of resources || []) {
    // add fields for backward compatibility
    const languageId = item.language;
    let [, resourceId] = (item.name || '').split(`${languageId}_`);
    resourceId = resourceId || item.name; // if language was not in name, then use name as resource ID
    item.resourceId = resourceId;
    item.languageId = languageId;
    item.checking_level = item.repo && item.repo.checking_level;
    item.foundInCatalog = 'NEW';
    item.modified = item.modified || item.released;

    if (item.zipball_url) {
      item.downloadUrl = item.zipball_url;
    }
    // check for version. if there is one, it will save having to fetch it from DCS later.
    if (item.release) { // if released
      const tagName = item.release.tag_name;
      if (tagName && (tagName[0] === 'v')) {
        item.version = tagName.substr(1);
      }
    } else {
      const branchOrTagName = item.branch_or_tag_name;
      if (branchOrTagName && (branchOrTagName[0] === 'v')) {
        item.version = branchOrTagName.substr(1);
      }
    }
    if (item.subject) {
      item.subject = item.subject.replaceAll(' ', '_');
    }
    // add supported resources to returned list
    if (item.downloadUrl && item.subject && item.name && item.full_name) {
      supported.push(item);
    }
  }
  return supported;
}

/**
 * get published catalog
 * @param {Object} searchParams - details below
 * @param {String} searchParams.owner - if undefined then all are searched
 * @param {String} searchParams.languageId - if undefined then all are searched
 * @param {String} searchParams.subject - one or more separated by comma.  If undefined then all are searched.
 *          Example `Bible,Aligned Bible,Greek New Testament,Hebrew Old Testament,Translation Words,TSV Translation Notes,Translation Academy'
 * @param {Number} searchParams.limit - maximum results to return, default 100
 * @param {Boolean} searchParams.partialMatch - if true will do case insensitive substring matching, default is false
 * @param {String} searchParams.stage - specifies which release stage to be returned out of these stages:
 *                    "prod" - return only the production releases (default)
 *                    "preprod" - return the pre-production release if it exists instead of the production release
 *                    "draft" - return the draft release if it exists instead of pre-production or production release
 *                   "latest" -return the default branch (e.g. master) if it is a valid RC instead of the "prod", "preprod" or "draft".  (default)
 * @param {Number} searchParams.checkingLevel - search only for entries with the given checking level(s). Can be 1, 2 or 3.  Default is any.
 * @param {Number} searchParams.sort - search only for entries with the given checking level(s). Can be 1, 2 or 3.  Default is any.
 * @param {number} retries - number of times to retry calling search API, default 3
 * @return {Promise<*[]|null>}
 */
export async function searchCatalogNext(searchParams, retries=3) {
  let result_ = null;
  const {
    owner,
    languageId,
    subject,
    limit = 100,
    stage = STAGE.LATEST,
    checkingLevel,
    partialMatch,
    sort = SORT.REPO_NAME,
  } = searchParams;

  try {
    let fetchUrl = `https://git.door43.org/api/catalog/v5/search`;
    let parameters = '';
    parameters = addUrlParameter(owner, parameters, 'owner');
    parameters = addUrlParameter(languageId, parameters, 'lang');
    parameters = addUrlParameter(subject, parameters, 'subject');
    parameters = addUrlParameter(limit, parameters, 'limit');
    parameters = addUrlParameter(stage, parameters, 'stage');
    parameters = addUrlParameter(checkingLevel, parameters, 'checkingLevel');
    parameters = addUrlParameter(partialMatch, parameters, 'partialMatch');
    parameters = addUrlParameter(sort, parameters, 'sort');
    if (parameters) {
      fetchUrl += '?' + parameters;
    }
    console.log(`Searching: ${fetchUrl}`);
    const {result} = await makeJsonRequestDetailed(fetchUrl, retries);
    result_ = result && result.data;
  } catch (e) {
    console.warn('searchCatalogNext() - error calling search API', e);
    return null;
  }

  const supported = getCompatibleResourceList(result_);
  return supported;
}

/**
 * does Catalog next API query to get manifest data
 * @param {string} owner
 * @param {string} repo
 * @param {number} retries
 * @return {Promise<{Object}>}
 */
export async function downloadManifestData(owner, repo, retries=5) {
  const fetchUrl = `https://git.door43.org/api/catalog/v5/entry/${owner}/${repo}/master/metadata`;
  try {
    const {result} = await makeJsonRequestDetailed(fetchUrl, retries);
    return result;
  } catch (e) {
    console.warn('getManifestData() - error getting manifest data', e);
    throw e;
  }
}
