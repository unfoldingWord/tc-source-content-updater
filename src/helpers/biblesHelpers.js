/* eslint-disable camelcase */
import fs from 'fs-extra';
import path from 'path-extra';

/**
 * generate manifest.json
 * @param {Object} oldManifest - old manifest data
 * @param {String} RESOURCE_OUTPUT_PATH - folder to store manifest.json
 * @param {Object} resource
 * @return {Object} new manifest data
 */
export function generateBibleManifest(oldManifest, RESOURCE_OUTPUT_PATH, resource) {
  const newManifest = {};
  newManifest.dublin_core = oldManifest.dublin_core; // preserve original manifest data
  newManifest.checking = oldManifest.checking;
  newManifest.projects = oldManifest.projects;
  newManifest.original_manifest = oldManifest;

  // copy some data for more convenient access
  newManifest.language_id = oldManifest.dublin_core.language.identifier;
  newManifest.language_name = oldManifest.dublin_core.language.title;
  newManifest.direction = oldManifest.dublin_core.language.direction;
  newManifest.subject = oldManifest.dublin_core.subject;
  newManifest.resource_id = oldManifest.dublin_core.identifier;
  newManifest.resource_title = oldManifest.dublin_core.title;
  const remoteModifiedTime = resource && resource.remoteModifiedTime || (new Date()).toJSON();
  newManifest.remoteModifiedTime = remoteModifiedTime;
  const oldMainfestIdentifier = oldManifest.dublin_core.identifier
                                      .toLowerCase();
  const identifiers = ['ugnt', 'ubh'];
  newManifest.description = identifiers.includes(oldMainfestIdentifier) ?
    'Original Language' : 'Gateway Language';

  const savePath = path.join(RESOURCE_OUTPUT_PATH, 'manifest.json');
  fs.writeFileSync(savePath, JSON.stringify(newManifest, null, 2));
  return newManifest;
}
