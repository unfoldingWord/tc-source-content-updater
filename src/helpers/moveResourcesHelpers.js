import fs from 'fs-extra';
import path from 'path';

/**
 * @description transfer an entire resource from source to target directory
 * @param {string} resourceSourcePath - current position of resource
 * @param {string} resourceTargetPath - folder where resources are moved
 * @return {Promise} Move directory Promise
 */
export function moveResources(resourceSourcePath, resourceTargetPath) {
  return new Promise((resolve, reject) => {
    if (resourceSourcePath && resourceSourcePath.length &&
            resourceTargetPath && resourceTargetPath.length &&
            fs.pathExistsSync(resourceSourcePath)) {
      if (fs.pathExistsSync(resourceTargetPath)) {
        fs.removeSync(resourceTargetPath);
      }
      fs.ensureDirSync(path.dirname(resourceTargetPath));
      fs.copySync(resourceSourcePath, resourceTargetPath);
      fs.removeSync(resourceSourcePath);
      resolve(resourceTargetPath);
    } else {
      reject('Invalid parameters to moveResources()');
    }
  });
}

