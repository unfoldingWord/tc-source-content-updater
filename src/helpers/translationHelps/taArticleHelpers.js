/* eslint-disable curly */
import { isObject } from 'util';
import fs from 'fs-extra';
import path from 'path-extra';
// heleprs
import * as resourcesHelpers from '../resourcesHelpers';
// constants
import * as errors from '../../resources/errors';

/**
 * @description Processes the extracted files for translationAcademy to create a single file for each
 * article
 * @param {Object} resource - Resource object
 * @param {String} sourcePath - Path to the extracted files that came from the zip file in the catalog
 * @param {String} outputPath - Path to place the processed files WITHOUT version in the path
 * @return {Boolean} true if success
 */
export function processTranslationAcademy(resource, sourcePath, outputPath) {
  if (!resource || !isObject(resource) || !resource.languageId || !resource.resourceId)
    throw Error(resourcesHelpers.formatError(resource, errors.RESOURCE_NOT_GIVEN));

  if (!sourcePath)
    throw Error(resourcesHelpers.formatError(resource, errors.SOURCE_PATH_NOT_GIVEN));

  if (!fs.pathExistsSync(sourcePath))
    throw Error(resourcesHelpers.formatError(resource, errors.SOURCE_PATH_NOT_EXIST));

  if (!outputPath)
    throw Error(resourcesHelpers.formatError(resource, errors.OUTPUT_PATH_NOT_GIVEN));

  if (fs.pathExistsSync(outputPath))
    fs.removeSync(outputPath);
  const manifest = resourcesHelpers.getResourceManifest(sourcePath);

  manifest.projects.forEach((project) => {
    const folderPath = path.join(sourcePath, project.path);
    const isDirectory = (item) => fs.lstatSync(path.join(folderPath, item)).isDirectory();
    const articleDirs = fs.readdirSync(folderPath).filter(isDirectory);

    articleDirs.forEach((articleDir) => {
      const titlePath = path.join(folderPath, articleDir, 'title.md');

      if (!fs.existsSync(titlePath)) {
        console.warn(`processTranslationAcademy() - title missing: ${titlePath}`);
        return;
      }

      let content = '# ' + fs.readFileSync(titlePath, 'utf8') + ' #\n';
      const articlePath = path.join(folderPath, articleDir, '01.md');

      if (!fs.existsSync(articlePath)) {
        console.warn(`processTranslationAcademy() - file missing: ${articlePath}`);
        return;
      }
      content += fs.readFileSync(articlePath, 'utf8');
      const destinationPath = path.join(
        outputPath,
        project.path,
        articleDir + '.md'
      );
      fs.outputFileSync(destinationPath, content);
    });
  });
  return true;
}
