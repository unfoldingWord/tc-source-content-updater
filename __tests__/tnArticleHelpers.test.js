/* eslint-disable quotes */
/* eslint-env jest */
import fs from 'fs-extra';
import path from 'path-extra';
import ospath from 'ospath';
// helpers
import * as tnArticleHelpers from '../src/helpers/translationHelps/tnArticleHelpers';
import {DOOR43_CATALOG} from '../src/helpers/apiHelpers';
import _ from "lodash";
import {parseUsfmOfBook} from '../src/helpers/packageParseHelpers';
import * as taArticleHelpers from '../src/helpers/translationHelps/taArticleHelpers';
import deepEqual from 'deep-equal';
// constants

const mockGetMissingOriginalResource = async (resourcesPath, originalLanguageId, originalLanguageBibleId, version, callLog) => {
  callLog.push({resourcesPath, originalLanguageId, originalLanguageBibleId, version}); // TRICKY - for ease of testing we are hijacking the DownloadErrors to keep track of calls
};

describe('Tests parseReference', function() {
  it('Test parseReference for test cases', async () => {
    const tests = [
      {ref: 'front:intro', expect: [{chapter: 'front', verse: 'intro'}]},
      {ref: '1:intro', expect: [{chapter: '1', verse: 'intro'}]},
      {ref: '1:1', expect: [{chapter: '1', verse: '1'}]},
      {ref: '1:1-2', expect: [{chapter: '1', verse: '1-2'}]},
      {ref: '1:1,3', expect: [{chapter: '1', verse: '1'}, {chapter: '1', verse: '3'}]},
      {ref: '1:1-2,4', expect: [{chapter: '1', verse: '1-2'}, {chapter: '1', verse: '4'}]},
      {ref: '1:1-2a,4', expect: [{chapter: '1', verse: '1-2'}, {chapter: '1', verse: '4'}]},
      {ref: '1:1b-2a,4', expect: [{chapter: '1', verse: '1-2'}, {chapter: '1', verse: '4'}]},
      {ref: '1:1-2,4b', expect: [{chapter: '1', verse: '1-2'}, {chapter: '1', verse: '4'}]},
      {ref: '1:1-2,4b,5-7a', expect: [{chapter: '1', verse: '1-2'}, {chapter: '1', verse: '4'}, {chapter: '1', verse: '5-7'}]},
      {ref: '1:1-2;2:4', expect: [{chapter: '1', verse: '1-2'}, {chapter: '2', verse: '4'}]},
      {ref: '1:1c-2b;2:4-5', expect: [{chapter: '1', verse: '1-2'}, {chapter: '2', verse: '4-5'}]},
    ];
    for (const test of tests) {
      const ref = test.ref;
      const expect_ = test.expect;
      const result = tnArticleHelpers.parseReference(ref);
      if (!deepEqual(result, expect_)) {
        console.log(`expect ${ref} to parse to ${JSON.stringify(expect_)}`);
        console.log(`  but got ${JSON.stringify(result)}`);
        expect(result).toEqual(expect_);
      }
    }
  });
});

describe('Tests tnArticleHelpers.processTranslationNotes()', function() {
  const resource = {
    languageId: 'en',
    resourceId: 'tn',
    version: '8',
    owner: DOOR43_CATALOG,
  };

  beforeEach(() => {
    fs.__resetMockFS();
  });

  it('Test tnArticleHelpers.processTranslationNotes() for en', async () => {
    // given
    const actualExtractedPath = path.join(__dirname, 'fixtures/en_tn-7-column');
    const mockedExtractedPath = path.join(ospath.home(), 'translationCore/resources/imports');
    fs.__loadDirIntoMockFs(actualExtractedPath, mockedExtractedPath);
    const resourcesPath = path.join(ospath.home(), 'translationCore/resources');
    const greekBiblePath = path.join(resourcesPath, 'el-x-koine', 'bibles/ugnt', 'v0.20_' + DOOR43_CATALOG);
    const greekSource = path.join(__dirname, 'fixtures/el-x-koine_ugnt');
    const greekTemp = path.join(ospath.home(), 'temp/grk');
    const bookFileName = '57-TIT.usfm';
    fs.__loadFilesIntoMockFs([bookFileName], greekSource, greekTemp);
    fs.ensureDirSync(greekBiblePath);
    parseUsfmOfBook(path.join(greekTemp, bookFileName), path.join(greekBiblePath, 'tit'));
    const hebrewBiblePath = path.join(resourcesPath, 'hbo', 'bibles/uhb', 'v2.1.19_' + DOOR43_CATALOG);
    fs.ensureDirSync(hebrewBiblePath);
    // copy greek to Hebrew as test filler
    fs.copySync(greekBiblePath, hebrewBiblePath);

    const resourcetHelpsPath = path.join(resourcesPath, resource.languageId, 'translationHelps');
    const tA_Path = path.join(resourcetHelpsPath, 'translationAcademy', 'v' + resource.version + '_' + DOOR43_CATALOG);
    const taSource = path.join(__dirname, 'fixtures/translationHelps/taExtractedFromCDN/en_ta');
    const taTemp = path.join(ospath.home(), 'temp/en_ta');
    fs.__loadDirIntoMockFs(taSource, taTemp);
    taArticleHelpers.processTranslationAcademy(resource, taTemp, tA_Path);

    const outputPath = path.join(resourcetHelpsPath, 'translationNotes', 'v' + resource.version + '_' + DOOR43_CATALOG);
    fs.ensureDirSync(outputPath);
    const expectedKtArticleListLength = 3;
    const expectedNamesArticleListLength = 2;
    const expectedIndexJson = [
      {id: 'apostle', name: 'apostle, apostles, apostleship'},
      {id: 'god', name: 'God'}, {id: 'sanctify', name: 'sanctify, sanctifies, sanctification'}
    ];
    const resource_ = _.cloneDeep(resource);
    const sourcePath = path.join(mockedExtractedPath, resource_.languageId + '_' + resource_.resourceId);
    const downloadErrors = [];
    const expectedTypes = [
      "discourse",
      "numbers",
      "figures",
      "culture",
      "grammar",
      "other"
    ];

    // when
    await tnArticleHelpers.processTranslationNotes(resource_, sourcePath, outputPath, resourcesPath, downloadErrors);
    const typeList = fs.readdirSync(outputPath);
    const grammarArticleList = fs.readJsonSync(path.join(outputPath, 'grammar/groups/tit/figs-exclusive.json'));

    // then
    expect(typeList).toEqual(expectedTypes);
    expect(grammarArticleList).toMatchSnapshot();
  }, 20000);
});

describe('Tests for tnArticleHelpers.getMissingResources()', function() {
  beforeEach(() => {
    fs.__resetMockFS();
  });

  it('Test for en with no dependencies', async () => {
    // given
    const fixturePath = path.join(__dirname, 'fixtures/translationHelps/tnExtractedFromCDN');
    const resourcesPath = path.join(ospath.home(), 'translationCore/resources');
    const importsPath = path.join(ospath.home(), resourcesPath, 'imports');
    fs.__loadDirIntoMockFs(fixturePath, importsPath);
    const sourcePath = path.join(importsPath, 'en_tn');
    const expectedCalls = [];
    const callLog = [];
    fs.writeJsonSync(path.join(resourcesPath, 'en', 'translationHelps/translationAcademy', 'v1_Door43-Catalog', 'manifest.json'), {}); // make dummy manifest for tA

    // when
    const {otQuery, ntQuery} = await tnArticleHelpers.getMissingResources(sourcePath, resourcesPath, mockGetMissingOriginalResource, callLog, 'en');

    // then
    validateCallLog(callLog, expectedCalls);
    expect(otQuery).toBeNull();
    expect(ntQuery).toBeNull();
  });

  it('Test for en with dependencies', async () => {
    // given
    const fixturePath = path.join(__dirname, 'fixtures/translationHelps/tnExtractedFromCDN');
    const resourcesPath = path.join(ospath.home(), 'translationCore/resources');
    const importsPath = path.join(ospath.home(), resourcesPath, 'imports');
    fs.__loadDirIntoMockFs(fixturePath, importsPath);
    const sourcePath = path.join(importsPath, 'en_tn');
    const manifestFixturePath = path.join(__dirname, 'fixtures/translationHelps/manifests/en_tn');
    fs.__loadFilesIntoMockFs(['manifest.yaml'], manifestFixturePath, sourcePath);
    const expectedCalls = [
      {
        "originalLanguageBibleId": "uhb",
        "originalLanguageId": "hbo",
        "resourcesPath": path.join(ospath.home(), 'translationCore/resources'),
        "version": "v2.1.9",
      },
      {
        "originalLanguageBibleId": "ugnt",
        "originalLanguageId": "el-x-koine",
        "resourcesPath": path.join(ospath.home(), 'translationCore/resources'),
        "version": "v0.10",
      },
    ];
    const callLog = [];
    const expectedNtQuery = '0.10';
    const expectedOtQuery = '2.1.9';
    fs.writeJsonSync(path.join(resourcesPath, 'en', 'translationHelps/translationAcademy', 'v1_Door43-Catalog', 'manifest.json'), {}); // make dummy manifest for tA

    // when
    const {otQuery, ntQuery} = await tnArticleHelpers.getMissingResources(sourcePath, resourcesPath, mockGetMissingOriginalResource, callLog, 'en');

    // then
    validateCallLog(callLog, expectedCalls);
    expect(ntQuery).toEqual(expectedNtQuery);
    expect(otQuery).toEqual(expectedOtQuery);
  });
});

//
// Helpers
//

/**
 * replace user specific paths in call log
 * @param {Array} callLog
 */
function cleanUpPaths(callLog) {
  const newLog = callLog.map((item) => {
    const newItem = {...item};
    if (item.resourcesPath) {
      const newPath = item.resourcesPath.replace(ospath.home(), '<HOME>');
      newItem.resourcesPath = newPath;
    }
    return newItem;
  });
  return newLog;
}

/**
 * validate call log against expected
 * @param {Array} callLog
 * @param {Array} expectedCalls
 */
function validateCallLog(callLog, expectedCalls) {
  const newCallLog = cleanUpPaths(callLog);
  const newExpectedCallLog = cleanUpPaths(expectedCalls);
  expect(newCallLog).toEqual(newExpectedCallLog);
}

