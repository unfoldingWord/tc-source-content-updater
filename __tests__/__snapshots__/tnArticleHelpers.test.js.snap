// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests tnArticleHelpers.processTranslationNotes() Test tnArticleHelpers.processTranslationNotes() for en 1`] = `
Array [
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "xy18",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "This includes Paul, Titus, and all Christians. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμῶν",
      "quoteString": "ἡμῶν",
      "reference": Object {
        "bookId": "tit",
        "chapter": 1,
        "verse": 3,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "xy17",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "This includes Paul, Titus, and all Christians. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμῶν",
      "quoteString": "ἡμῶν",
      "reference": Object {
        "bookId": "tit",
        "chapter": 1,
        "verse": 4,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "xy15",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "This includes Paul, Titus, and all Christians. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμῶν",
      "quoteString": "ἡμῶν",
      "reference": Object {
        "bookId": "tit",
        "chapter": 2,
        "verse": 8,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "xy16",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "Here, **our** is inclusive of Paul, Titus, and all Christians (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμῶν",
      "quoteString": "ἡμῶν",
      "reference": Object {
        "bookId": "tit",
        "chapter": 2,
        "verse": 10,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "abce",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "This includes Paul, Titus, and all Christians. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμᾶς",
      "quoteString": "ἡμᾶς",
      "reference": Object {
        "bookId": "tit",
        "chapter": 2,
        "verse": 12,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "xy20",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "This includes Paul, Titus, and all Christians. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμῶν",
      "quoteString": "ἡμῶν",
      "reference": Object {
        "bookId": "tit",
        "chapter": 2,
        "verse": 14,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "xy21",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "This includes Paul, Titus, and all Christians. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμᾶς",
      "quoteString": "ἡμᾶς",
      "reference": Object {
        "bookId": "tit",
        "chapter": 2,
        "verse": 14,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "bl8e",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "“even we” or “we ourselves” This includes Paul, Titus, and all Christians, referring to the time before they trusted in Christ. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμεῖς",
      "quoteString": "ἡμεῖς",
      "reference": Object {
        "bookId": "tit",
        "chapter": 3,
        "verse": 3,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "abch",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "This includes Paul, Titus, and all Christians. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμῶν",
      "quoteString": "ἡμῶν",
      "reference": Object {
        "bookId": "tit",
        "chapter": 3,
        "verse": 4,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "xy24",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "This includes Paul, Titus, and all Christians. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμᾶς",
      "quoteString": "ἡμᾶς",
      "reference": Object {
        "bookId": "tit",
        "chapter": 3,
        "verse": 6,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "xy23",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "This includes Paul, Titus, and all Christians. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμῶν",
      "quoteString": "ἡμῶν",
      "reference": Object {
        "bookId": "tit",
        "chapter": 3,
        "verse": 6,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "xy33",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "Here, **our** includes Paul and Titus. The form should be either dual or inclusive. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": Array [
        Object {
          "occurrence": 1,
          "word": "οἱ",
        },
        Object {
          "occurrence": 1,
          "word": "ἡμέτεροι",
        },
      ],
      "quoteString": "οἱ ἡμέτεροι",
      "reference": Object {
        "bookId": "tit",
        "chapter": 3,
        "verse": 14,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
  Object {
    "comments": false,
    "contextId": Object {
      "checkId": "xy35",
      "glQuote": "",
      "groupId": "figs-exclusive",
      "occurrence": 1,
      "occurrenceNote": "Here, **us** is probably exclusive and refers to Paul and the group of Christians with him. Paul is sending greetings from this group to the group of Christians that is with Titus on Crete. (See: [[rc://*/ta/man/translate/figs-exclusive]])",
      "quote": "ἡμᾶς",
      "quoteString": "ἡμᾶς",
      "reference": Object {
        "bookId": "tit",
        "chapter": 3,
        "verse": 15,
      },
      "tool": "translationNotes",
    },
    "nothingToSelect": false,
    "reminders": false,
    "selections": false,
    "verseEdits": false,
  },
]
`;
